src/istio/mixerclient/client_impl.cc

```diff
@@ -170,82 +191,136 @@ CancelFunc MixerClientImpl::Check(CheckContextSharedPtr &context,
+                                  const TransportCheckFunc &transport,
+                                  const CheckDoneFunc &on_done) {
+  //
+  // This lambda and any lambdas it creates for retry will inc the ref count
+  // on the CheckContext shared pointer.
+  //
+  // The other captures (this/MixerClientImpl, transport func, and on_done func
+  // are assumed to be long-lived objects that exist until shutdown.
+  //
+  CancelFunc cancel_func = transport(
+      context->request(), context->response(),
+      [this, context, transport, on_done](const Status &status) {
+        context->resetCancel();
+
+        //
+        // Classify and track transport errors
+        //
+
+        TransportResult result = TransportStatus(status);
+
+        switch (result) {
+          case TransportResult::SUCCESS:
+            ++total_remote_call_successes_;
+            break;
+          case TransportResult::RESPONSE_TIMEOUT:
+            ++total_remote_call_timeouts_;
+            break;
+          case TransportResult::SEND_ERROR:
+            ++total_remote_call_send_errors_;
+            break;
+          case TransportResult::OTHER:
+            ++total_remote_call_other_errors_;
+            break;
+        }
+
+        if (result != TransportResult::SUCCESS && context->retryable()) {
+          ++total_remote_call_retries_;
+          const uint32_t retry_ms = RetryDelay(context->retryAttempt());
+
+          MIXER_DEBUG("Retry %u in %u msec due to transport error=%s",
+                      context->retryAttempt() + 1, retry_ms,
+                      status.ToString().c_str());
+
+          context->retry(retry_ms,
+                         timer_create_([this, context, transport, on_done]() {
+                           RemoteCheck(context, transport, on_done);
+                         }));
+
+          return;
+        }
+
+        //
+        // Update caches.  This has the side-effect of updating
+        // status, so track those too
+        //
+
+        if (!context->policyCacheHit()) {
+          context->updatePolicyCache(status, *context->response());
+
+          if (context->policyStatus().ok()) {
+            ++total_remote_check_accepts_;
+          } else {
+            ++total_remote_check_denies_;
+          }
+        }
+
+        if (context->quotaCheckRequired()) {
+          context->updateQuotaCache(status, *context->response());
+
+          if (context->quotaStatus().ok()) {
+            ++total_remote_quota_accepts_;
+          } else {
+            ++total_remote_quota_denies_;
+          }
+        }
+
+        MIXER_DEBUG(
+            "CheckResult transport=%s, policy=%s, quota=%s, attempt=%u",
+            status.ToString().c_str(),
+            result == TransportResult::SUCCESS
+                ? context->policyStatus().ToString().c_str()
+                : "NA",
+            result == TransportResult::SUCCESS && context->quotaCheckRequired()
+                ? context->policyStatus().ToString().c_str()
```

Comment:
I'm impressed you're reading this so thoroughly.  Thank you for the diligent code review :)

```
  // Return a combination of the error code name and message.
  string ToString() const;
```

So it returns a copy that isn't stored in a variable... then I call c_str() on it.   Will it be destroyed before or after I call c_str()?

Interestingly, because the log threshold is warn in these tests, asan won't actually see this.

@jplevyak do you know what will happen here?